modules/mixins/cacheMixin.js

```diff
@@ -363,41 +379,64 @@
-                      if (model) {
-                        models.push(model);
-                      }
+          return $.ajax(ajaxOpts)
+            .done(function(data) {
+              var i, requesterIdx, requesterIdsAsDict, models, privateCollection,
+                  requesterLength, requesters, model,
+                  requestedIdsLength = requestedIds.length,
+                  setOptions = options.setOptions;
+              collection.set(collection.parse(data), setOptions);
+              // Set respective collection's models for requested ids only.
+              requesters = collection.getRequesters();
+              requesterLength = requesters.length;
+              // For each requester...
+              for (requesterIdx = 0; requesterIdx < requesterLength; requesterIdx++) {
+                requesterIdsAsDict = collection.getRequesterIdsAsDictionary(requesters[requesterIdx]);
+                models = [];
+                // ... now let's iterate over the ids that were fetched ...
+                for (i = 0; i < requestedIdsLength; i++) {
+                  //if the id that the requester cares about matches that model whose id was just fetched...
+                  if (requesterIdsAsDict[requestedIds[i]]) {
+                    model = collection.get(requestedIds[i]);
+                    // if the model was removed, no worries, the parent won't attempt to update the child on that one.
+                    if (model) {
+                      models.push(model);
                     }
                   }
-                  privateCollection = collection.knownPrivateCollections[requesters[requesterIdx]];
-                  // a fetch by the parent will not remove a model in a requester collection that wasn't fetched with this call
-                  privateCollection.set(models, {remove: false});
                 }
-              });
+                privateCollection = collection.knownPrivateCollections[requesters[requesterIdx]];
+                // a fetch by the parent will not remove a model in a requester collection that wasn't fetched with this call
+                privateCollection.set(models, {remove: false});
+              }
+            });
         }, options)
-        .always(function() {
-          delete collection.idsFetching[fetchId];
-        });
+          .always(function() {
+            fetchComplete = true;
+            delete collection.idsFetching[fetchId];
+            _.each(requestedIds, function(requestedId) {
+              if (collection.idPromises) {
+                var existingPromisesForId = collection.idPromises[requestedId];
+                collection.idPromises[requestedId] = _.without(existingPromisesForId, fetchPromise);
+              }
+            });
+          });
+
+        if (!collection.idPromises) {
+          collection.idPromises = {};
+        }
+        if (!fetchComplete) { // fixes the sync case (mainly during tests when mockjax is used).
```

Comment:
just the if (!fetchComplete) { is for the test, the internals would be run every time otherwise.